// uppy/lib/index.js
'use strict';

var Core = require('./core/index.js'

// Parent
);var Plugin = require('./plugins/Plugin'

// Orchestrators
);var Dashboard = require('./plugins/Dashboard/index.js'

// Acquirers
);var Dummy = require('./plugins/Dummy');
var DragDrop = require('./plugins/DragDrop/index.js');
var FileInput = require('./plugins/FileInput.js');
var GoogleDrive = require('./plugins/GoogleDrive/index.js');
var Dropbox = require('./plugins/Dropbox/index.js');
var Instagram = require('./plugins/Instagram/index.js');
var Webcam = require('./plugins/Webcam/index.js'

// Progressindicators
);var StatusBar = require('./plugins/StatusBar');
var ProgressBar = require('./plugins/ProgressBar.js');
var Informer = require('./plugins/Informer.js'

// Modifiers
);var MetaData = require('./plugins/MetaData.js'

// Uploaders
);var Tus10 = require('./plugins/Tus10');
var XHRUpload = require('./plugins/XHRUpload');
var Transloadit = require('./plugins/Transloadit');
var AwsS3 = require('./plugins/AwsS3'

// Other?
);var RestoreFiles = require('./plugins/RestoreFiles');

module.exports = {
  Core: Core,
  Plugin: Plugin,
  Dummy: Dummy,
  StatusBar: StatusBar,
  ProgressBar: ProgressBar,
  Informer: Informer,
  DragDrop: DragDrop,
  GoogleDrive: GoogleDrive,
  Dropbox: Dropbox,
  Instagram: Instagram,
  FileInput: FileInput,
  Tus10: Tus10,
  XHRUpload: XHRUpload,
  Transloadit: Transloadit,
  AwsS3: AwsS3,
  Dashboard: Dashboard,
  MetaData: MetaData,
  Webcam: Webcam,
  RestoreFiles: RestoreFiles
};
